WordPress/wordcamp.org
# Tooling: Changes to composer and docker

I was frustrated by having the 3rd-party plugins and themes included in our composer.json file, even though they're only necessary for the Docker environment. It felt cumbersome to work around them, since as a committer, I have the svn repo, and don't need them to be installed by composer anyway. So I set out to split the plugins and themes into a separate composer.json file that only runs within the Docker container, as discussed in #159. Along the way, I had to significantly refactor the provisioning scripts used by Docker. I also discovered some unrelated problems with some of our apt-get packages in the php-fpm container, so I had to fix those to get the environment to function. So here's a list of all the changes:

* Remove 3rd-party plugins and themes from main composer.json file.
* Add a new composer.json file with only plugins and themes, that is only used during Docker container provisioning.
* Update docker configs to install from the new composer.json file during `docker-compose up`, but only if wp-content does not contain an .svn directory.
* Replace apt-get packages installed during docker provisioning that are no longer available (mysql-client and ssmtp)
* Add setup and config for the alternate smtp package (msmtp).
* Update the readme for Docker.

Fixes #159 

**To test with the svn repo**

* Go through the setup steps in .docker/readme.md. You might already have it all set up, but you'll probably need to re-run `composer update` from the project root.
* Run `docker-compose up --build` (without the `-d`). This will show you the output coming from within the containers after they have finished building. You should see a line that says "Starting up..." followed by one that says "The wp-content directory appears to be a Subversion repository. Skipping additional setup...". This means it didn't try to install the 3rd-party plugins and themes via composer.

**To test without the svn repo**

* Clone a fresh copy of this repo somewhere.
* Follow the instructions in .docker/readme.md.
* Run `docker-compose up --build` (without the `-d`). This will show you the output coming from within the containers after they have finished building. You should see a line that says "Starting up..." followed by several lines from composer as it installs 3rd-party plugins and themes.

**To test MailCatcher with the new msmtp package**

* Using either setup from above, with the containers up and running, open the MailCatcher dashboard in a new tab: `http://localhost:1080/`
* Get WP to send an email. One way to do this is via `https://central.wordcamp.test/wp-admin/post.php?post=2935&action=edit&classic-editor`. Check the box to "Manually send this email" and pick one of the test WordCamps from the dropdown. Then click Update.
* Back in the MailCatcher dashboard, the email should appear.
